diff --git a/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aic8800d80n_compat.c b/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aic8800d80n_compat.c
index 0d2d753..cc4e279 100644
--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aic8800d80n_compat.c
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aic8800d80n_compat.c
@@ -1,3 +1,4 @@
+#include <linux/vmalloc.h>
 #include "aic8800d80n_compat.h"
 #include "aic_bsp_export.h"

diff --git a/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aic8800dc_compat.c b/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aic8800dc_compat.c
index 64fbb01..7768b54 100644
--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aic8800dc_compat.c
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aic8800dc_compat.c
@@ -1,4 +1,5 @@
 #include <linux/version.h>
+#include <linux/vmalloc.h>
 #include "aic8800dc_compat.h"
 #include "aic_bsp_driver.h"

diff --git a/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aic_bsp_driver.c b/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aic_bsp_driver.c
index f723a22..eaf0274 100644
--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aic_bsp_driver.c
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aic_bsp_driver.c
@@ -476,7 +476,7 @@ void rwnx_rx_handle_msg(struct aic_sdio_dev *sdiodev, struct ipc_e2a_msg *msg)
 }

 #if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 4, 0)
-MODULE_IMPORT_NS(VFS_internal_I_am_really_a_filesystem_and_am_NOT_a_driver);
+MODULE_IMPORT_NS("VFS_internal_I_am_really_a_filesystem_and_am_NOT_a_driver");
 #endif

 #define MD5(x) x[0],x[1],x[2],x[3],x[4],x[5],x[6],x[7],x[8],x[9],x[10],x[11],x[12],x[13],x[14],x[15]
diff --git a/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aicsdio.c b/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aicsdio.c
index 83f45cb..d86e461 100644
--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aicsdio.c
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aicsdio.c
@@ -1461,7 +1461,7 @@ int aicwf_sdio_busrx_thread(void *data)
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 14, 0)
 static void aicwf_sdio_bus_pwrctl(struct timer_list *t)
 {
-	struct aic_sdio_dev *sdiodev = from_timer(sdiodev, t, timer);
+	struct aic_sdio_dev *sdiodev = container_of(t, typeof(*sdiodev), timer);
 #else
 static void aicwf_sdio_bus_pwrctl(ulong data)
 {
@@ -1666,7 +1666,7 @@ void aicwf_sdio_pwrctl_timer(struct aic_sdio_dev *sdiodev, uint duration)
 	spin_lock_bh(&sdiodev->pwrctl_lock);
 	if (!duration) {
 		if (timer_pending(&sdiodev->timer))
-			del_timer_sync(&sdiodev->timer);
+			timer_delete_sync(&sdiodev->timer);
 	} else {
 		sdiodev->active_duration = duration;
 		timeout = msecs_to_jiffies(sdiodev->active_duration);
diff --git a/src/SDIO/driver_fw/driver/aic8800/aic8800_btlpm/aic8800_btlpm.c b/src/SDIO/driver_fw/driver/aic8800/aic8800_btlpm/aic8800_btlpm.c
index 52d768d..93410be 100644
--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_btlpm/aic8800_btlpm.c
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_btlpm/aic8800_btlpm.c
@@ -579,7 +579,7 @@ static int bluesleep_start(void)

 	return 0;
 fail:
-	del_timer(&rx_timer);
+	timer_delete(&rx_timer);
 	atomic_inc(&open_count);

 	return retval;
@@ -602,7 +602,7 @@ static void bluesleep_stop(void)
 	/* assert BT_WAKE */
 	gpio_set_value(bsi->ext_wake, bsi->ext_wake_assert);

-	del_timer(&rx_timer);
+	timer_delete(&rx_timer);
 	clear_bit(BT_PROTO, &flags);

 	if (test_bit(BT_ASLEEP, &flags)) {
@@ -949,7 +949,7 @@ static int bluesleep_remove(struct platform_device *pdev)
 		if (disable_irq_wake(bsi->host_wake_irq))
 			BT_ERR("Couldn't disable hostwake IRQ wakeup mode\n");
 		free_irq(bsi->host_wake_irq, &bsi->pdev->dev);
-		del_timer(&rx_timer);
+		timer_delete(&rx_timer);
 		if (test_bit(BT_ASLEEP, &flags))
 			hsuart_power(1);
 	}
diff --git a/src/SDIO/driver_fw/driver/aic8800/aic8800_btlpm/lpm.c b/src/SDIO/driver_fw/driver/aic8800/aic8800_btlpm/lpm.c
index 6e9b715..120f88b 100644
--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_btlpm/lpm.c
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_btlpm/lpm.c
@@ -558,7 +558,7 @@ static int bluesleep_start(void)

 	return 0;
 fail:
-	del_timer(&rx_timer);
+	timer_delete(&rx_timer);
 	atomic_inc(&open_count);

 	return retval;
@@ -581,7 +581,7 @@ static void bluesleep_stop(void)
 	/* assert BT_WAKE */
 	gpio_set_value(bsi->ext_wake, bsi->ext_wake_assert);

-	del_timer(&rx_timer);
+	timer_delete(&rx_timer);
 	clear_bit(BT_PROTO, &flags);

 	if (test_bit(BT_ASLEEP, &flags)) {
@@ -933,7 +933,7 @@ static int bluesleep_remove(struct platform_device *pdev)
 		if (disable_irq_wake(bsi->host_wake_irq))
 			BT_ERR("Couldn't disable hostwake IRQ wakeup mode\n");
 		free_irq(bsi->host_wake_irq, &bsi->pdev->dev);
-		del_timer(&rx_timer);
+		timer_delete(&rx_timer);
 		if (test_bit(BT_ASLEEP, &flags))
 			hsuart_power(1);
 	}
diff --git a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/aic_priv_cmd.c b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/aic_priv_cmd.c
index 69c0475..1448ba8 100644
--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/aic_priv_cmd.c
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/aic_priv_cmd.c
@@ -11,6 +11,7 @@
  */

 #include <linux/netdevice.h>
+#include <linux/vmalloc.h>
 #include <linux/ctype.h>
 #include "rwnx_defs.h"
 #include "rwnx_msg_tx.h"
diff --git a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/aicwf_sdio.c b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/aicwf_sdio.c
index bbd54a0..45fa57c 100644
--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/aicwf_sdio.c
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/aicwf_sdio.c
@@ -260,7 +260,7 @@ static void aicwf_netif_timer(struct timer_list *t)
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4, 14, 0)
 		struct aic_sdio_dev *sdiodev = (struct aic_sdio_dev *) data;
 #else
-		struct aic_sdio_dev *sdiodev = from_timer(sdiodev, t, netif_timer);
+		struct aic_sdio_dev *sdiodev = container_of(t, typeof(*sdiodev), netif_timer);
 #endif

 	if (!work_pending(&sdiodev->netif_work))
@@ -296,7 +296,7 @@ static void aicwf_temp_ctrl_timer(struct timer_list *t)
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4, 14, 0)
 	struct aic_sdio_dev *sdiodev = (struct aic_sdio_dev *) data;
 #else
-	struct aic_sdio_dev *sdiodev = from_timer(sdiodev, t, tp_ctrl_timer);
+	struct aic_sdio_dev *sdiodev = container_of(t, typeof(*sdiodev), tp_ctrl_timer);
 #endif

 	if (!work_pending(&sdiodev->tp_ctrl_work))
@@ -1022,12 +1022,12 @@ static int aicwf_sdio_suspend(struct device *dev)
 	}

 #ifdef CONFIG_TEMP_CONTROL
-	del_timer_sync(&sdiodev->tp_ctrl_timer);
+	timer_delete_sync(&sdiodev->tp_ctrl_timer);
 	cancel_work_sync(&sdiodev->tp_ctrl_work);

 	mod_timer(&sdiodev->tp_ctrl_timer, jiffies + msecs_to_jiffies(TEMP_GET_INTERVAL));

-	del_timer_sync(&sdiodev->netif_timer);
+	timer_delete_sync(&sdiodev->netif_timer);
 	cancel_work_sync(&sdiodev->netif_work);
 #endif

@@ -1280,13 +1280,13 @@ void aicwf_sdio_exit(void)
 #ifdef CONFIG_TEMP_CONTROL
 		if (timer_pending(&g_rwnx_plat->sdiodev->tp_ctrl_timer)) {
 			AICWFDBG(LOGINFO, "%s del tp_ctrl_timer\n", __func__);
-			del_timer_sync(&g_rwnx_plat->sdiodev->tp_ctrl_timer);
+			timer_delete_sync(&g_rwnx_plat->sdiodev->tp_ctrl_timer);
 		}
 		cancel_work_sync(&g_rwnx_plat->sdiodev->tp_ctrl_work);

 		if (timer_pending(&g_rwnx_plat->sdiodev->netif_timer)) {
 			AICWFDBG(LOGINFO, "%s del netif_timer\n", __func__);
-			del_timer_sync(&g_rwnx_plat->sdiodev->netif_timer);
+			timer_delete_sync(&g_rwnx_plat->sdiodev->netif_timer);
 		}
 		cancel_work_sync(&g_rwnx_plat->sdiodev->netif_work);
 #endif
@@ -2804,7 +2804,7 @@ static void aicwf_sdio_bus_pwrctl(struct timer_list *t)
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4, 14, 0)
 	struct aic_sdio_dev *sdiodev = (struct aic_sdio_dev *) data;
 #else
-	struct aic_sdio_dev *sdiodev = from_timer(sdiodev, t, timer);
+	struct aic_sdio_dev *sdiodev = container_of(t, typeof(*sdiodev), timer);
 #endif

 	if (sdiodev->bus_if->state == BUS_DOWN_ST) {
@@ -3001,7 +3001,7 @@ void aicwf_sdio_pwrctl_timer(struct aic_sdio_dev *sdiodev, uint duration)
 	spin_lock_bh(&sdiodev->pwrctl_lock);
 	if (!duration) {
 		if (timer_pending(&sdiodev->timer))
-			del_timer_sync(&sdiodev->timer);
+			timer_delete_sync(&sdiodev->timer);
 	} else {
 		sdiodev->active_duration = duration;
 		timeout = msecs_to_jiffies(sdiodev->active_duration);
diff --git a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/aicwf_tcp_ack.c b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/aicwf_tcp_ack.c
index ca47b26..dc77902 100644
--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/aicwf_tcp_ack.c
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/aicwf_tcp_ack.c
@@ -106,7 +106,7 @@ void tcp_ack_deinit(struct rwnx_hw *priv)
 		drop_msg = NULL;

 		write_seqlock_bh(&ack_m->ack_info[i].seqlock);
-		del_timer(&ack_m->ack_info[i].timer);
+		timer_delete(&ack_m->ack_info[i].timer);
 		drop_msg = ack_m->ack_info[i].msgbuf;
 		ack_m->ack_info[i].msgbuf = NULL;
 		write_sequnlock_bh(&ack_m->ack_info[i].seqlock);
@@ -375,7 +375,7 @@ int tcp_ack_handle(struct msg_buf *new_msgbuf,
 				//printk("%lx \n",ack_info->msgbuf);
 				drop_msg = ack_info->msgbuf;
 				ack_info->msgbuf = NULL;
-				del_timer(&ack_info->timer);
+				timer_delete(&ack_info->timer);
 			}else{
 				//printk("msgbuf is NULL \n");
 			}
@@ -409,7 +409,7 @@ int tcp_ack_handle(struct msg_buf *new_msgbuf,
 				   atomic_read(&ack_m->max_drop_cnt)))) {
 			ack_info->drop_cnt = 0;
 			ack_info->in_send_msg = new_msgbuf;
-			del_timer(&ack_info->timer);
+			timer_delete(&ack_info->timer);
 		} else {
 			ret = 1;
 			ack_info->msgbuf = new_msgbuf;
@@ -472,7 +472,7 @@ int tcp_ack_handle_new(struct msg_buf *new_msgbuf,
 			ack_info->drop_cnt = 0;
 			//send_msg = new_msgbuf;
 			ack_info->in_send_msg = new_msgbuf;
-			del_timer(&ack_info->timer);
+			timer_delete(&ack_info->timer);
 		}else{
 			ret = 1;
 			ack_info->msgbuf = new_msgbuf;
diff --git a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_gki.c b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_gki.c
index 7f6333d..c2e8e19 100644
--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_gki.c
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_gki.c
@@ -48,7 +48,7 @@ static bool __rwnx_cfg80211_unexpected_frame(struct net_device *dev, u8 cmd,
 }

 bool rwnx_cfg80211_rx_spurious_frame(struct net_device *dev,
-				const u8 *addr, gfp_t gfp)
+				const u8 *addr, int link_id, gfp_t gfp)
 {
 	struct wireless_dev *wdev = dev->ieee80211_ptr;
 	bool ret;
@@ -63,7 +63,7 @@ bool rwnx_cfg80211_rx_spurious_frame(struct net_device *dev,
 }

 bool rwnx_cfg80211_rx_unexpected_4addr_frame(struct net_device *dev,
-				const u8 *addr, gfp_t gfp)
+				const u8 *addr, int link_id, gfp_t gfp)
 {
 	struct wireless_dev *wdev = dev->ieee80211_ptr;
 	bool ret;
diff --git a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_gki.h b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_gki.h
index a41f578..96213b3 100644
--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_gki.h
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_gki.h
@@ -10,10 +10,10 @@


 bool rwnx_cfg80211_rx_spurious_frame(struct net_device *dev,
-				const u8 *addr, gfp_t gfp);
+				const u8 *addr, int link_id, gfp_t gfp);

 bool rwnx_cfg80211_rx_unexpected_4addr_frame(struct net_device *dev,
-				const u8 *addr, gfp_t gfp);
+				const u8 *addr, int link_id, gfp_t gfp);

 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0))
 void rwnx_cfg80211_notify_new_peer_candidate(struct net_device *dev, const u8 *addr,
diff --git a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_main.c b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_main.c
index 1e2e73d..b977cf2 100644
--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_main.c
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_main.c
@@ -1766,7 +1766,7 @@ void aicwf_p2p_alive_timeout(struct timer_list *t)
 	rwnx_vif = (struct rwnx_vif *)data;
 	rwnx_hw = rwnx_vif->rwnx_hw;
 	#else
-	rwnx_hw = from_timer(rwnx_hw, t, p2p_alive_timer);
+	rwnx_hw = container_of(t, typeof(*rwnx_hw), p2p_alive_timer);
 	rwnx_vif = rwnx_hw->p2p_dev_vif;
 	#endif

@@ -2203,7 +2203,7 @@ static void rwnx_cfgp2p_stop_p2p_device(struct wiphy *wiphy, struct wireless_dev
 	if (rwnx_vif == rwnx_hw->p2p_dev_vif) {
 		rwnx_hw->is_p2p_alive = 0;
 		if (timer_pending(&rwnx_hw->p2p_alive_timer)) {
-			del_timer_sync(&rwnx_hw->p2p_alive_timer);
+			timer_delete_sync(&rwnx_hw->p2p_alive_timer);
 		}

 		if (rwnx_vif->up) {
@@ -3559,7 +3559,7 @@ static int rwnx_cfg80211_stop_ap(struct wiphy *wiphy, struct net_device *dev)
 	rwnx_vif->ap.start = false;
 	aicwf_nl_hook_deinit(rwnx_vif->ap.band, rwnx_vif->rwnx_hw->iface_idx);
 	if (timer_pending(&rwnx_vif->steer_timer))
-		del_timer_sync(&rwnx_vif->steer_timer);
+		timer_delete_sync(&rwnx_vif->steer_timer);
 	cancel_work_sync(&rwnx_vif->steer_work);
 #if 0
 	flush_workqueue(rwnx_vif->rsp_wq);
@@ -3599,6 +3599,7 @@ static int rwnx_cfg80211_stop_ap(struct wiphy *wiphy, struct net_device *dev)
  * configured at firmware level.
  */
 static int rwnx_cfg80211_set_monitor_channel(struct wiphy *wiphy,
+											 struct net_device *dev,
 											 struct cfg80211_chan_def *chandef)
 {
 	struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
@@ -3653,7 +3654,7 @@ static int rwnx_cfg80211_set_monitor_channel(struct wiphy *wiphy,

 int rwnx_cfg80211_set_monitor_channel_(struct wiphy *wiphy,
                                              struct cfg80211_chan_def *chandef){
-    return rwnx_cfg80211_set_monitor_channel(wiphy, chandef);
+    return rwnx_cfg80211_set_monitor_channel(wiphy, NULL, chandef);
 }


@@ -3710,7 +3711,7 @@ void rwnx_cfg80211_mgmt_frame_register(struct wiphy *wiphy,
  *	have changed. The actual parameter values are available in
  *	struct wiphy. If returning an error, no value should be changed.
  */
-static int rwnx_cfg80211_set_wiphy_params(struct wiphy *wiphy, u32 changed)
+static int rwnx_cfg80211_set_wiphy_params(struct wiphy *wiphy, int radio_idx, u32 changed)
 {
 	return 0;
 }
@@ -3724,6 +3725,7 @@ static int rwnx_cfg80211_set_wiphy_params(struct wiphy *wiphy, u32 changed)
  *	(as advertised by the nl80211 feature flag.)
  */
 static int rwnx_cfg80211_set_tx_power(struct wiphy *wiphy, struct wireless_dev *wdev,
+									  int radio_idx,
 									  enum nl80211_tx_power_setting type, int mbm)
 {
 	struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
@@ -3752,14 +3754,10 @@ static int rwnx_cfg80211_set_tx_power(struct wiphy *wiphy, struct wireless_dev *
 }

 static int rwnx_cfg80211_get_tx_power(struct wiphy *wiphy,
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 8, 0)
  struct wireless_dev *wdev,
-#endif
+	int radio_idx, unsigned int link_id,
 	int *mbm)
 {
-    #if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
-    struct wireless_dev *wdev = NULL;
-    #endif
     //struct rwnx_hw *rwnx_hw = wiphy_priv(wiphy);
     //struct rwnx_vif *vif;
     s8 pwr = 0;
@@ -4072,7 +4070,7 @@ static int rwnx_cfg80211_get_channel(struct wiphy *wiphy,

 	if (rwnx_vif->vif_index == rwnx_hw->monitor_vif) {
 		//retrieve channel from firmware
-		rwnx_cfg80211_set_monitor_channel(wiphy, NULL);
+		rwnx_cfg80211_set_monitor_channel(wiphy, NULL, NULL);
 	}

 	//Check if channel context is valid
@@ -5738,7 +5736,7 @@ static void aicwf_pwrloss_timer(struct timer_list *t)
 	rwnx_vif = (struct rwnx_vif *)data;
 	rwnx_hw = rwnx_vif->rwnx_hw;
 #else
-	rwnx_hw = from_timer(rwnx_hw, t, pwrloss_timer);
+	rwnx_hw = container_of(t, typeof(*rwnx_hw), pwrloss_timer);
 #endif
 	if (!work_pending(&rwnx_hw->pwrloss_work))
 		schedule_work(&rwnx_hw->pwrloss_work);
@@ -6315,7 +6313,7 @@ void rwnx_cfg80211_deinit(struct rwnx_hw *rwnx_hw)

 #ifdef CONFIG_DYNAMIC_PWR
 	if(timer_pending(&rwnx_hw->pwrloss_timer)){
-		del_timer_sync(&rwnx_hw->pwrloss_timer);}
+		timer_delete_sync(&rwnx_hw->pwrloss_timer);}
 	cancel_work_sync(&rwnx_hw->pwrloss_work);
 #endif

@@ -6324,7 +6322,7 @@ void rwnx_cfg80211_deinit(struct rwnx_hw *rwnx_hw)
 		list_for_each_entry(defrag_ctrl, &rwnx_hw->defrag_list, list) {
 			list_del_init(&defrag_ctrl->list);
 			if (timer_pending(&defrag_ctrl->defrag_timer))
-				del_timer_sync(&defrag_ctrl->defrag_timer);
+				timer_delete_sync(&defrag_ctrl->defrag_timer);
 			dev_kfree_skb(defrag_ctrl->skb);
 			kfree(defrag_ctrl);
 		}
@@ -6456,7 +6454,7 @@ static void __exit rwnx_mod_exit(void)


 #if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 4, 0)
-MODULE_IMPORT_NS(VFS_internal_I_am_really_a_filesystem_and_am_NOT_a_driver);
+MODULE_IMPORT_NS("VFS_internal_I_am_really_a_filesystem_and_am_NOT_a_driver");
 #endif

 module_init(rwnx_mod_init);
diff --git a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_msg_tx.c b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_msg_tx.c
index 7435c74..66d5035 100644
--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_msg_tx.c
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_msg_tx.c
@@ -10,6 +10,7 @@
  ******************************************************************************
  */

+#include <linux/vmalloc.h>
 #include "rwnx_msg_tx.h"
 #include "rwnx_mod_params.h"
 #include "reg_access.h"
@@ -308,7 +309,7 @@ static int rwnx_send_msg(struct rwnx_hw *rwnx_hw, const void *msg_params,

 	//RWNX_DBG(RWNX_FN_ENTRY_STR);
     AICWFDBG(LOGTRACE, "%s (%d)%s reqcfm:%d in_irq:%d in_softirq:%d in_atomic:%d\r\n",
-    __func__, reqid, RWNX_ID2STR(reqid), reqcfm, (int)in_irq(), (int)in_softirq(), (int)in_atomic());
+    __func__, reqid, RWNX_ID2STR(reqid), reqcfm, (int)in_hardirq(), (int)in_softirq(), (int)in_atomic());


 #ifdef AICWF_USB_SUPPORT
@@ -416,7 +417,7 @@ static int rwnx_send_msg1(struct rwnx_hw *rwnx_hw, const void *msg_params,

 //	RWNX_DBG(RWNX_FN_ENTRY_STR);
     printk("%s (%d)%s reqcfm:%d in_irq:%d in_softirq:%d in_atomic:%d\r\n",
-    __func__, reqid, RWNX_ID2STR(reqid), reqcfm, (int)in_irq(), (int)in_softirq(), (int)in_atomic());
+    __func__, reqid, RWNX_ID2STR(reqid), reqcfm, (int)in_hardirq(), (int)in_softirq(), (int)in_atomic());

 	rwnx_wakeup_lock(rwnx_hw->ws_tx);
 	msg = container_of((void *)msg_params, struct lmac_msg, param);
diff --git a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_platform.c b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_platform.c
index cc1b479..9497101 100644
--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_platform.c
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_platform.c
@@ -548,7 +548,7 @@ static int rwnx_plat_tl4_fw_upload(struct rwnx_plat *rwnx_plat, u8 *fw_addr,
 #endif

 #if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 4, 0)
-MODULE_IMPORT_NS(VFS_internal_I_am_really_a_filesystem_and_am_NOT_a_driver);
+MODULE_IMPORT_NS("VFS_internal_I_am_really_a_filesystem_and_am_NOT_a_driver");
 #endif

 #if 0
diff --git a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_rx.c b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_rx.c
index 2348323..332d117 100644
--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_rx.c
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_rx.c
@@ -1659,7 +1659,7 @@ int reord_flush_tid(struct aicwf_rx_priv *rx_priv, struct sk_buff *skb, u8 tid)
 	preorder_ctrl->enable = false;
 	spin_unlock_irqrestore(&preorder_ctrl->reord_list_lock, flags);
 	if (timer_pending(&preorder_ctrl->reord_timer))
-		ret = del_timer_sync(&preorder_ctrl->reord_timer);
+		ret = timer_delete_sync(&preorder_ctrl->reord_timer);
 	cancel_work_sync(&preorder_ctrl->reord_timer_work);

 	return 0;
@@ -1684,7 +1684,7 @@ void reord_deinit_sta(struct aicwf_rx_priv *rx_priv, struct reord_ctrl_info *reo
 		if(preorder_ctrl->enable){
 			preorder_ctrl->enable = false;
 			if (timer_pending(&preorder_ctrl->reord_timer)) {
-				ret = del_timer_sync(&preorder_ctrl->reord_timer);
+				ret = timer_delete_sync(&preorder_ctrl->reord_timer);
 			}
 			cancel_work_sync(&preorder_ctrl->reord_timer_work);
 		}
@@ -1895,7 +1895,7 @@ void reord_timeout_handler (struct timer_list *t)
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4, 14, 0)
 	struct reord_ctrl *preorder_ctrl = (struct reord_ctrl *)data;
 #else
-	struct reord_ctrl *preorder_ctrl = from_timer(preorder_ctrl, t, reord_timer);
+	struct reord_ctrl *preorder_ctrl = container_of(t, typeof(*preorder_ctrl), reord_timer);
 #endif

 #if 0 //AIDEN
@@ -2044,7 +2044,7 @@ int reord_process_unit(struct aicwf_rx_priv *rx_priv, struct sk_buff *skb, u16 s
 		}
 	} else {
 	if (timer_pending(&preorder_ctrl->reord_timer)) {
-			ret = del_timer(&preorder_ctrl->reord_timer);
+			ret = timer_delete(&preorder_ctrl->reord_timer);
 	}
 	}
 	
@@ -2146,7 +2146,7 @@ void defrag_timeout_cb(struct timer_list *t)
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4, 15, 0)
 	defrag_ctrl = (struct defrag_ctrl_info *)data;
 #else
-	defrag_ctrl = from_timer(defrag_ctrl, t, defrag_timer);
+	defrag_ctrl = container_of(t, typeof(*defrag_ctrl), defrag_timer);
 #endif

 	printk("%s:%p\r\n", __func__, defrag_ctrl);
@@ -2374,9 +2374,9 @@ check_len_update:
 		rwnx_vif = rwnx_rx_get_vif(rwnx_hw, hw_rxhdr->flags_vif_idx);
 		if (rwnx_vif) {
 #ifdef CONFIG_GKI
-			rwnx_cfg80211_rx_spurious_frame(rwnx_vif->ndev, hdr->addr2, GFP_ATOMIC);
+			rwnx_cfg80211_rx_spurious_frame(rwnx_vif->ndev, hdr->addr2, -1, GFP_ATOMIC);
 #else
-            cfg80211_rx_spurious_frame(rwnx_vif->ndev, hdr->addr2, GFP_ATOMIC);
+            cfg80211_rx_spurious_frame(rwnx_vif->ndev, hdr->addr2, -1, GFP_ATOMIC);
 #endif
 		}
 		goto end;
@@ -2695,7 +2695,7 @@ check_len_update:
 							skb_tmp = defrag_info->skb;
 							list_del_init(&defrag_info->list);
 							if (timer_pending(&defrag_info->defrag_timer)) {
-								ret = del_timer(&defrag_info->defrag_timer);
+								ret = timer_delete(&defrag_info->defrag_timer);
 							}
 							kfree(defrag_info);
 							spin_unlock_bh(&rwnx_hw->defrag_lock);
@@ -2749,10 +2749,10 @@ check_len_update:
 					printk("aicwf: 4addr flag error\n");
 #ifdef CONFIG_GKI
 					rwnx_cfg80211_rx_unexpected_4addr_frame(rwnx_vif->ndev,
-													   sta->mac_addr, GFP_ATOMIC);
+													   sta->mac_addr, -1, GFP_ATOMIC);
 #else
                     cfg80211_rx_unexpected_4addr_frame(rwnx_vif->ndev,
-                                                       sta->mac_addr, GFP_ATOMIC);
+                                                       sta->mac_addr, -1, GFP_ATOMIC);
 #endif
 				}
 			}
diff --git a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_wakelock.c b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_wakelock.c
index 1d15296..edc39db 100644
--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_wakelock.c
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_wakelock.c
@@ -1,28 +1,20 @@
 #include <linux/kernel.h>
 #include <linux/version.h>
 #include <linux/platform_device.h>
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 2, 0)
-#include <linux/pm_wakeirq.h>
-#else
 #include <linux/pm_wakeup.h>
-#endif
 #include "rwnx_defs.h"
 #include "rwnx_wakelock.h"

 struct wakeup_source *rwnx_wakeup_init(const char *name)
 {
-	struct wakeup_source *ws;
-	ws = wakeup_source_create(name);
-	wakeup_source_add(ws);
-	return ws;
+	return wakeup_source_register(NULL, name);
 }

 void rwnx_wakeup_deinit(struct wakeup_source *ws)
 {
 	if (ws && ws->active)
 		__pm_relax(ws);
-	wakeup_source_remove(ws);
-	wakeup_source_destroy(ws);
+	wakeup_source_unregister(ws);
 }

 struct wakeup_source *rwnx_wakeup_register(struct device *dev, const char *name)
